snowplow-tracker/src/androidTest/java/com/snowplowanalytics/snowplow/tracker/LinkDecoratorTest.kt

```diff
@@ -0,0 +1,199 @@
+package com.snowplowanalytics.snowplow.tracker
+
+
+import android.net.Uri
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.test.platform.app.InstrumentationRegistry
+import com.snowplowanalytics.core.tracker.CrossDeviceParameter
+import com.snowplowanalytics.core.utils.Util.base64Encode
+import com.snowplowanalytics.snowplow.Snowplow
+import com.snowplowanalytics.snowplow.configuration.NetworkConfiguration
+import com.snowplowanalytics.snowplow.configuration.SessionConfiguration
+import com.snowplowanalytics.snowplow.configuration.SubjectConfiguration
+import com.snowplowanalytics.snowplow.configuration.TrackerConfiguration
+import com.snowplowanalytics.snowplow.controller.SessionController
+import com.snowplowanalytics.snowplow.controller.TrackerController
+import com.snowplowanalytics.snowplow.event.ScreenView
+import com.snowplowanalytics.snowplow.network.HttpMethod
+import com.snowplowanalytics.snowplow.util.TimeMeasure
+import org.junit.Assert
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import java.util.concurrent.TimeUnit
+
+
+@RunWith(AndroidJUnit4::class)
+class LinkDecoratorTest {
+    private lateinit var tracker: TrackerController
+    private lateinit var session: SessionController
+    private lateinit var userId: String
+    private lateinit var appId: String
+    private val subjectUserId = "subjectUserId"
+    private val subjectUserIdEncoded = Uri.encode(base64Encode(subjectUserId))
+    private val testLink = Uri.parse("http://example.com")
+    private fun matchesRegex(pattern: Regex, result: Uri) {
+        Assert.assertTrue(
+            "$result\ndoes not match expected: $pattern",
+            pattern.matches(result.toString())
+        )
+    }
+
+
+    @Before
+    fun before() {
+        tracker = getTracker()
+        session = tracker.session!!
+        userId = session.userId
+        appId = Uri.encode(base64Encode(tracker.appId))
+    }
+
+    @Test
+    fun testWithoutSession() {
+        val tracker = getTrackerNoSession()
+        val result = tracker.decorateLink(testLink)
+        Assert.assertEquals(null, result)
+    }
+
+    @Test
+    fun testDecorateUriWithReason() {
+        tracker.track(ScreenView("test"))
+
+        val reason = "reasonTest"
+        val reasonEncoded = Uri.encode(base64Encode(reason))
+
+        val pattern =
+            Regex("""http://example\.com\?_sp=$userId\.\d{13}\.${session.sessionId}\.\.$appId\.\.$reasonEncoded""")
+        val result = tracker.decorateLink(testLink, reason = reason)
+
+        matchesRegex(pattern, result!!)
+    }
+
+    @Test
+    fun testDecorateUriWithNoParams() {
+        tracker.track(ScreenView("test"))
+
+        val pattern =
+            Regex("""http://example\.com\?_sp=$userId\.\d{13}\.${session.sessionId}\.\.$appId""")
```

Comment:
```suggestion
            Regex("""^http://example\.com\?_sp=$userId\.\d{13}\.${session.sessionId}\.\.$appId$""")
```

Nit: I think these regular expressions are a bit tricky because they don't test what comes after the appId, so we don't test that the trailing dots are trimmed. Would be nicer to add the `$` at the end to test that as well.